README.md
muppet
======

Python implementation of Mutual, although rather preliminary


CHANGES
* 0.1.0 (in development)
    * Initial version of muppet

test/test_remote_channel.py
import time
from muppet.remote_channel import RemoteChannel

class TestRemoteChannel:

  def setup_class(self):
    self.redis_options = {"redis": {"host": "127.0.0.1", "port": 6379}}
    self.sender = None
    self.receiver = None
    self.calledbackChannel = ""
    self.calledbackMessage = ""

  def __callback(self, channel, message):
    self.calledbackChannel = channel
    self.calledbackMessage = message
    self.receiver.end()

  def test_messaging(self):
    self.sender = RemoteChannel("greeting", self.redis_options)
    self.receiver = RemoteChannel("greeting", self.redis_options)

    # listen for messages
    self.receiver.listen(self.__callback)

    # send message
    self.sender.send("hello")
    self.sender.end()

    # wait for timeout
    time.sleep(2)

    # assert to check we received the message
    assert self.calledbackChannel == "greeting"
    assert self.calledbackMessage == "hello"

test/test_durable_channel.py
import time
from muppet.durable_channel import DurableChannel

class TestDurableChannel:

  def setup_class(self):
    self.redis_options = {"redis": {"host": "127.0.0.1", "port": 6379}}
    self.dispatcher = None
    self.worker = None
    self.calledbackMessage = None

  def __timeout_callback(self, message):
    self.calledbackMessage = message
    self.worker.end()
    self.dispatcher.end()

  def test_messaging(self):
    self.dispatcher = DurableChannel("dispatcher.1", self.redis_options)
    self.worker = DurableChannel("worker.1", self.redis_options)

    # dispatch message
    self.dispatcher.send(content="task", to="worker.1")

    # receive message
    message = self.worker.receive()
    assert message["content"] == "task"
    self.worker.reply(message=message, response="reply", timeout=5000)

    # receive reply
    reply = self.dispatcher.receive()
    assert reply["content"] == "reply"
    self.dispatcher.close(reply)
    self.worker.end()
    self.worker = None
    self.dispatcher.end()
    self.dispatcher = None

  def test_timeout(self):
    self.dispatcher = DurableChannel("dispatcher.2", self.redis_options, self.__timeout_callback)
    self.worker = DurableChannel("worker.2", self.redis_options)

    # dispatch message
    self.dispatcher.send(content="task", to="worker.2", timeout=1000)

    # receive message
    message = self.worker.receive()
    assert message["content"] == "task"
    
    # wait for timeout
    time.sleep(2)

    # assert to check message timed out 
    assert self.calledbackMessage["content"] == "task"

muppet/remote_channel.py
import string
import random
import json
import threading
import redis

class RemoteChannel:

  def __init__(self, name, options):
    if name == None or len(name) == 0:
      raise ValueError("Remote channels cannot be anonymous")

    self.name = name
    self.id = ''.join(random.choice(string.letters + string.digits) for _ in xrange(16))
    self.redis = redis.StrictRedis(host=options["redis"]["host"], port=options["redis"]["port"], db=0)
    self.pubsub = self.redis.pubsub()
    self.listening = False

  def __startMessageLoop(self):
    for message in self.pubsub.listen():
      if message["type"] == "message":
        if message["data"] == self.id + "__kill__":
          self.pubsub.unsubscribe()
          break
        elif not message["data"].endswith("__kill__") :
          data = json.loads(message["data"])
          self.callback(message["channel"], data["content"])

  def send(self, content):
    message = {"content": content}
    self.redis.publish(self.name, json.dumps(message))

  def listen(self, callback):
    if not self.listening:
      self.listening = True
      self.callback = callback
      self.pubsub.subscribe([self.name])
      self.listenerThread = threading.Thread(target=self.__startMessageLoop)
      self.listenerThread.daemon = True
      self.listenerThread.start()

  def end(self):
    if self.listening:
      self.redis.publish(self.name, self.id + "__kill__")

muppet/durable_channel.py
import string
import random
import json
import time
import threading
import redis

class DurableChannel:

  def __init__(self, name, options, timeoutCallback=None):
    if name == None or len(name) == 0:
      raise ValueError("Durable channels cannot be anonymous")

    self.name = name
    self.stopping = False
    self.redis = redis.StrictRedis(host=options["redis"]["host"], port=options["redis"]["port"], db=0)
    self.timeoutCallback = timeoutCallback
    self.timeoutMonitorFrequency = 1000
    if "timeoutMonitorFrequency" in options:
      self.timeoutMonitorFrequency = options["timeoutMonitorFrequency"]
    
    self.timeoutMonitorThread = threading.Thread(target=self.__monitorTimeouts)
    self.timeoutMonitorThread.daemon = True
    self.timeoutMonitorThread.start()

  def __get_current_time(self):
    return int(round(time.time() * 1000))

  def __package(self, content, to, request_id=None, timeout=None):
    newId = ''.join(random.choice(string.letters + string.digits) for _ in xrange(16))
    message = {
      "id": newId,
      "requestId": request_id,
      "from": self.name,
      "to": to,
      "timeout": timeout,
      "content": content
    }
    return message

  def __monitorTimeouts(self):
    while True:
      if self.stopping:
        break

      time.sleep(self.timeoutMonitorFrequency / 1000)

      expiredMessageIds = self.redis.zrangebyscore(
        self.name + ".pending", 
        0, 
        self.__get_current_time()
      )
      if len(expiredMessageIds) == 0:
        continue

      for expiredMessageId in expiredMessageIds:
        expiredMessageTokens = expiredMessageId.split("::")

        expiredMessage = self.redis.hget(
          expiredMessageTokens[0] + ".messages", 
          expiredMessageTokens[1]
        )

        self.redis.zrem(
          self.name + ".pending", 
          expiredMessageId
        )

        if expiredMessage == None:
          continue

        expiredMessage = json.loads(expiredMessage)
        
        self.redis.hdel(
          expiredMessageTokens[0] + ".messages", 
          expiredMessageTokens[1]
        )

        if self.timeoutCallback != None:
          _message = {"content": expiredMessage["content"]}
          if "requestId" in expiredMessage and expiredMessage["requestId"] != None:
            _message["requestId"] = expiredMessage["requestId"]
          self.timeoutCallback(_message)

  def send(self, content, to, timeout=None):
    message = self.__package(
      content=content, 
      to=to, 
      timeout=timeout
    )    

    self.redis.hset(
      to + ".messages", 
      message["id"], 
      json.dumps(message)
    )

    if timeout != None:
      self.redis.zadd(
        self.name + ".pending", 
        (self.__get_current_time() + timeout), 
        to + "::" + message["id"]
      )

    self.redis.lpush(
      to + ".queue", 
      message["id"]
    )

  def receive(self):
    message = None
    while message == None:
      messageId = self.redis.brpop(
        self.name + ".queue"
      )
      messageId = messageId[1]

      message = self.redis.hget(
        self.name + ".messages", 
        messageId
      )
    
    message = json.loads(message)

    if "requestId" in message and message["requestId"] != None:
      if message["timeout"] != None:
        self.redis.zrem(
          self.name + ".pending", 
          message["from"] + "::" + message["requestId"]
        )
      self.redis.hdel(
        message["from"] + ".messages", 
        message["requestId"]
      )
    
    _message = {"content": message["content"]}
    if "requestId" in message and message["requestId"] != None:
      _message["from"] = message["to"]
      _message["to"] = message["from"]
      _message["requestId"] = message["requestId"]
      _message["responseId"] = message["id"]
    else:
      _message["from"] = message["from"]
      _message["to"] = message["to"]
      _message["requestId"] = message["id"]

    return _message

  def reply(self, message, response, timeout=None):
    request = self.redis.hget(
      self.name + ".messages", 
      message["requestId"]
    )
    request = json.loads(request)
    message = self.__package(
      content=response, 
      to=request["from"], 
      request_id=message["requestId"], 
      timeout=timeout
    )

    self.redis.hset(
      request["from"] + ".messages", 
      message["id"], 
      json.dumps(message)
    )

    if timeout != None:
      self.redis.zadd(
        self.name + ".pending", 
        (self.__get_current_time() + timeout), 
        request["from"] + "::" + message["id"]
      )

    self.redis.lpush(
      request["from"] + ".queue", 
      message["id"]
    )

  def close(self, message):
    self.redis.hdel(
      self.name + ".messages", 
      message["responseId"]
    )
    self.redis.zrem(
      message["to"] + ".pending", 
      message["from"] + "::" + message["responseId"]
    )

  def end(self):
    self.stopping = True